server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java

```diff
@@ -122,12 +146,40 @@ public UpdateMessagePatch build() {
         return isAnswered;
     }
 
-    public boolean isFlagsIdentity() {
+    public Optional<Set<JmapFlag>> getKeywords() {
+        return keywords;
+    }
+
+    public boolean isFlagsIdentity(Flags currentFlags) {
+        Preconditions.checkState(noConflict(), "Does not support both is* and keyword");
+        if (keywords.isPresent()) {
+            return noKeywordOnFlags(currentFlags)
+                && noFlagOnKeywords(ImmutableSet.copyOf(currentFlags.getSystemFlags()));
+        }
         return !isAnswered.isPresent()
             && !isFlagged.isPresent()
             && !isUnread.isPresent();
     }
 
+    private boolean noKeywordOnFlags(Flags currentFlags) {
```

Comment:
From earlier remarks, I believe this code is not needed, but it should be written in a functional way...